odoo/orm/models.py

```diff
@@ -6104,28 +6104,86 @@ def build_like_regex(value: str, exact: bool):
+            # sort records by name in descending order, then by id
+            records.sorted('name DESC, id')
+            # sort records using default order
+            records.sorted()
         """
-        if key is None:
-            if any(self._ids):
-                ids = self.search([('id', 'in', self.ids)])._ids
-            else:  # Don't support new ids because search() doesn't work on new records
-                ids = self._ids
-            ids = tuple(reversed(ids)) if reverse else ids
-        else:
-            if isinstance(key, str):
-                key = itemgetter(key)
-            ids = tuple(item.id for item in sorted(self, key=key, reverse=reverse))
+        if isinstance(key, str):
+            key = self._sorted_order_to_function(key)
+        elif key is None:
+            key = self._sorted_order_to_function(self._order)
+        ids = tuple(item.id for item in sorted(self, key=key, reverse=reverse))
         return self.__class__(self.env, ids, self._prefetch_ids)
 
+    @api.model
+    def _sorted_order_to_function(self, order: str) -> Callable[[BaseModel], tuple]:
+        def order_to_function(order_part):
+            order_match = regex_order.match(order_part)
+            if not order_match:
+                raise ValueError(f"Invalid order {order!r} to sort")
+            field_name = order_match['field']
+            property_name = order_match['property']
+            reverse = (order_match['direction'] or '').upper() == 'DESC'
+            nulls = (order_match['nulls'] or '').upper()
+            if nulls:
+                nulls_first = nulls == 'NULLS FIRST'
+            else:
+                nulls_first = reverse
+
+            field = self._fields[field_name]
+            if field.type == 'many2one' and (not property_name or property_name == 'id'):
+                seen = self.env.context.get('__m2o_order_seen', ())
+                if field in seen:
+                    return lambda _: None
+                comodel = self.env[field.comodel_name].with_context(__m2o_order_seen=frozenset((field, *seen)))
+                func_comodel = comodel._sorted_order_to_function(property_name or comodel._order)
+
+                def getter(rec):
+                    value = rec[field_name]
+                    if not value:
+                        return None
+                    return func_comodel(value)
+            elif field.relational:
+                raise ValueError(f"Invalid order on relational field {order_part!r} to sort")
+            else:
+                def getter(rec):
+                    value = field.__get__(rec)
+                    if value is False and field.type != 'boolean':
+                        return None
+                    if isinstance(value, NewId):
+                        return value.origin
+                    if property_name:
+                        # XXX property_name not correct (for dates); wait for getter in odoo-dev:master-domain-filtered-krma
```

Comment:
I will check this after my holidays and after filtered_domain is based on `Domain` (the PR introduces getters on fields.